public/app/features/variables/state/onTimeRangeUpdated.test.ts

```diff
@@ -449,8 +469,97 @@ describe('when onTimeRangeUpdated is dispatched', () => {
-      expect(spyTimeRangeUpdated).toHaveBeenCalledWith(expectedVariables.interval);
-      expect(spyTimeRangeUpdated).toHaveBeenCalledTimes(2);
+      expect(spyTimeRangeUpdated).toHaveBeenCalledWith(expectedVariables.queryD);
+      expect(spyTimeRangeUpdated).toHaveBeenCalledTimes(3);
+
+      spyTimeRangeUpdated.mockRestore();
+    });
+
+    // query variables can depend on datasource variables, but currently datasource variables do not refresh on time
+    // range change via the UI. This test ensures that query variables that depend on datasource variables are refreshed
+    it('Should ensure query variable is refreshed when dependent on a non-refreshing datasource variable', async () => {
+      const dataSourceVariable = datasourceBuilder()
+        .withId('dsVar')
+        .withRootStateKey('key')
+        .withName('datasource-1')
+        .withOptions('a ds')
+        .withCurrent('a ds')
+        .build();
+
+      const queryE = queryBuilder()
+        .withId('e')
+        .withName('e')
+        .withRootStateKey('key')
+        .withQuery('query ${dsVar}')
+        .withRefresh(VariableRefresh.onTimeRangeChanged)
+        .build();
+
+      const queryF = queryBuilder()
+        .withId('f')
+        .withName('f')
+        .withRootStateKey('key')
+        .withQuery('query ${e}')
+        .withRefresh(VariableRefresh.onTimeRangeChanged)
+        .build();
+
+      const customVariables = {
+        dsVar: dataSourceVariable,
+        queryE,
+        queryF,
+      };
+
+      const { key, preloadedState, range, dependencies } = getTestContextVariables(
+        getDashboardModel(),
+        customVariables
+      );
+
+      // Spying on timeRangeUpdated action
+      const spyTimeRangeUpdated = jest.spyOn(actions, 'timeRangeUpdated');
+
+      // Initiating the Redux testing environment
+      await reduxTester<RootReducerType>({ preloadedState, debug: true })
+        .givenRootReducer(getRootReducer())
+        .whenActionIsDispatched(toKeyedAction(key, variablesInitTransaction({ uid: key })))
+        .whenAsyncActionIsDispatched(onTimeRangeUpdated(key, range, dependencies), true);
+
+      // Defining the variables that are expected to be refreshed
+      // Based on the algorithm of getVariablesThatNeedRefreshNew, we have the following variables:
+      // - queryA, queryD and interval: These are variables with no dependents. According to the algorithm, should be refreshed.
+      // - queryE: This is a query variable with a dependent node (queryF) but also has a dependency on DsVar datasource variable.
+      //   because DSVar is not configured to refresh on time range,this means that queryE is the
+      //   variable that is expected to be refreshed
+      // Note: Variables 'b' and 'c', f are not expected to be in the refreshed variables list even though they're set to
+      // refresh on time range change. This is because they have dependencies ('b' and 'c' depend on 'a').
```

Comment:
You are describing queryA and queryD variables, but these are not defined in this test. Are those shared across multiple tests?

Reply:
indeed, I have the function `getTestContextVariables` that sets some base variables, and in this PR I added an extra variable D (with no dependents) to account for the edge case I missed last time :)